design-proposals/klip-28-create-or-replace

```diff
@@ -0,0 +1,156 @@
+
+## Design
+
+If the `source_name` does not yet exist, a `CREATE OR REPLACE` statement functions identically to
+a normal `CREATE` statement. Otherwise, ksqlDB executes the following:
+
+1. Identify the original `queryID` that populates the source (`INSERT INTO` discussed later)
+2. Ensure the upgrade is valid
+3. Terminate `queryID`
+4. Start the new query under the same `queryID`
+
+A few changes need to happen in order to make this work. For 1, we need to maintain a mapping from
+source to queryID(s). If the source has multiple associated ids (in the case of `INSERT INTO`) then
+the upgrade will fail and not terminate any queries.
+
+For step 2, there will be a component to determine whether two topologies are "upgrade
+compatible"; the first iterations, which will be delivered as part of this KLIP, will only allow
+for the most basic upgrades: 
+
+- Any _transparent_ upgrade will be supported
+- Any _data selection_ upgrade will be supported
+- _Schema evolution_ upgrades will be supported on simple and stateful queries, but it will be
+    communicated that the users will not get `backfill` or `ordered` properties for stateful.
+- _Source modifying_ upgrades will not be supported
+- _Topology changes_ will not be supported
+
+There are currently discussions that discuss how to expand the support of some of these upgrades,
+but we believe there is value in supporting the limitted set described above.
+
+For step 4, we need to be able to generate queryIDs differently for `CREATE OR REPLACE` statements
+than for others (i.e. it shouldn't just be the offset of the command, but rather the same queryID 
+as the original query.) One simple way to implement this is to allow the queryID to be specified
+in the command topic. Since the engine receiving the request has a complete view of the engine
+metadata, it can determine the queryID to enqueue onto the command topic.
+
+### On "INSERT INTO"
+
+Insert into can eventually be replaced with `UNION` as proposed in [KLIP-17](https://github.com/confluentinc/ksql/pull/4125),
+but that must happen in lock-step with this proposal. At first, we will support `CREATE OR REPLACE`
+and `INSERT INTO`. Then, we will add support for `UNION`, allowing us to model consecutive `INSERT INTO`
+statements as replacing unions with larger unions (essentially adding an extra source to the union).
+The approach for that will require a slight modification to the four steps outlined in the design
+section above.
+
+## Test plan
+
+_What are the failure scenarios you are going to cover in your testing? What scale testing do you plan to run? What about peformance and load testing? It goes 
+without saying that most classes should have unit tests._
+
+## Documentation Updates
+
+We will add the following `md` file:
+
+### Synopsis:
+
+```sql
+CREATE OR REPLACE (STREAM | TABLE) source_name 
+    [WITH ( property_name = expression [, ...] )]
+    AS query;
+```
```

Comment:
Rather than add a new file, can we not just add the `OR REPLACE` bit to the current `CREATE TABLE AS` and `CREATE STREAM AS` documentation?

Reply:
+1